File: Task/Tau-function/Java/tau-function.java
public class TauFunction {
    private static long divisorCount(long n) {
        long total = 1;
        // Deal with powers of 2 first
        for (; (n & 1) == 0; n >>= 1) {
            ++total;
        }
        // Odd prime factors up to the square root
        for (long p = 3; p * p <= n; p += 2) {
            long count = 1;
            for (; n % p == 0; n /= p) {
                ++count;
            }
            total *= count;
        }
        // If n > 1 then it's prime
        if (n > 1) {
            total *= 2;
        }
        return total;
    }

    public static void main(String[] args) {
        final int limit = 100;
        System.out.printf("Count of divisors for the first %d positive integers:\n", limit);
        for (long n = 1; n <= limit; ++n) {
            System.out.printf("%3d", divisorCount(n));
            if (n % 20 == 0) {
                System.out.println();
            }
        }
    }
}

File: Task/Tau-function/Scala/tau-function.scala
object TauFunction {

  private def divisorCount(n: Long): Long = {
    var count = 1L
    var number = n

    // Deal with powers of 2 first
    while ((number & 1L) == 0) {
      count += 1
      number >>= 1
    }

    // Odd prime factors up to the square root
    var p = 3L
    while (p * p <= number) {
      var tempCount = 1L
      while (number % p == 0) {
        tempCount += 1
        number /= p
      }
      count *= tempCount
      p += 2
    }

    // If n > 1 then it's prime
    if (number > 1) {
      count *= 2
    }

    count
  }

  def main(args: Array[String]): Unit = {
    val limit = 100
    println(s"Count of divisors for the first $limit positive integers:")
    for (n <- 1 to limit) {
      print(f"${divisorCount(n)}%3d")
      if (n % 20 == 0) println()
    }
  }
}
